### PR TITLE
Implement SELinux policy for QubesDB

### PR DESCRIPTION
**Pull request by DemiMarie**
This allows using QubesDB with SELinux enforcing.  Part of QubesOS/qubes-issues#4239.

### CODE REVIEW

File: rpm_spec/qubes-db-vm.spec.in [hunk 1]
```diff
@@ -41,18 +43,80 @@ BuildRequires: systemd
 
 Source0: qubes-db-%{version}.tar.gz
 
+%bcond_without selinux
+%if %{with selinux}
+Requires: (%{name}-selinux if selinux-policy)
+%package selinux
+
+# The following is based on %%selinux_requires, which has the following license:
```
Comment (marmarek) [inline rpm_spec/qubes-db-vm.spec.in]: Same comment as in https://github.com/QubesOS/qubes-linux-utils/pull/80

File: selinux/qubes-core-qubesdb.te [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+	allow qubes_qubesdb_daemon_t qubes_var_run_t:dir { add_name remove_name search write };
+}
+
+ipp_qubesdb_talk(xdm_t)
+ipp_qubesdb_talk(xend_t)
+ipp_qubesdb_talk(NetworkManager_t)
+ipp_qubesdb_talk(local_login_t)
+ipp_qubesdb_talk(pulseaudio_t)
+ipp_qubesdb_talk(staff_t)
+ipp_qubesdb_talk(sysadm_t)
```
Comment (marmarek) [inline selinux/qubes-core-qubesdb.te:32]: I think `udev_t` needs to be here too (for scripts publishing block and usb devices to qubesdb).

**Review by marmarek — CHANGES_REQUESTED** [review]

File: rpm_spec/qubes-db-vm.spec.in [hunk 3]
```diff
@@ -41,18 +43,57 @@ BuildRequires: systemd
 
 Source0: qubes-db-%{version}.tar.gz
 
+%bcond_without selinux
+%if %{with selinux}
+Requires: (%{name}-selinux if selinux-policy)
+%package selinux
+
+%selinux_requires
```
Comment (marmarek) [inline rpm_spec/qubes-db-vm.spec.in]: Same issue as in linux-utils

**Review by marmarek — CHANGES_REQUESTED** [review]

**Comment by marmarek** [comment]
Same as in gui-agent. Take a look at CI logs. Check CI status after you push subsequent fix.

**Comment by marmarek** [comment]
qubesdb fails to start:
```
avc:  denied  { write } for  pid=227 comm="qubesdb-daemon" name="qubes" dev="tmpfs" ino=71 scontext=system_u:system_r:qubes_qubesdb_daemon_t:s0 tcontext=system_u:object_r:var_run_t:s0 tclass=dir permissive=0

```

**Comment by DemiMarie** [comment]
Problem looks like wrong SELinux context on `/var/run/qubes`.  Most recent force-push should fix it.

**Comment by DemiMarie** [comment]
PipelineRetry

**Comment by marmarek** [comment]
PipelineRefresh

**Comment by marmarek** [comment]
```
DEBUG: make: Entering directory '/builddir/build/BUILD/qubes-db-4.1.13/selinux'
DEBUG: Compiling targeted qubes-core-qubesdb module
DEBUG: qubes-core-qubesdb.te:20:ERROR 'unknown type var_run_t' at token ';' on line 3904:
DEBUG:  	
DEBUG: 	type_transition qubes_qubesdb_daemon_t var_run_t:dir qubes_var_run_t "qubes";
DEBUG: /usr/bin/checkmodule:  error(s) encountered while parsing configuration
```

**Comment by marmarek** [comment]
```
[2022-07-18 04:06:59]          Starting [0;1;39mqubes-db.service[0m - Qubes DB agent...
[2022-07-18 04:06:59] [   20.363165] systemd[1]: qubes-db.service: Control process exited, code=exited, status=1/FAILURE
[2022-07-18 04:06:59] [   20.379767] systemd[1]: qubes-db.service: Failed with result 'exit-code'.
[2022-07-18 04:06:59] [   20.404639] systemd[1]: Failed to start qubes-db.service - Qubes DB agent.
```

This happens with selinux not even enabled yet (just updated package installed).

**Comment by DemiMarie** [comment]
> ```
> [2022-07-18 04:06:59]          Starting �[0;1;39mqubes-db.service�[0m - Qubes DB agent...
> [2022-07-18 04:06:59] [   20.363165] systemd[1]: qubes-db.service: Control process exited, code=exited, status=1/FAILURE
> [2022-07-18 04:06:59] [   20.379767] systemd[1]: qubes-db.service: Failed with result 'exit-code'.
> [2022-07-18 04:06:59] [   20.404639] systemd[1]: Failed to start qubes-db.service - Qubes DB agent.
> ```
> 
> This happens with selinux not even enabled yet (just updated package installed).

Ouch, and no error message either.

I think the best thing to do is to have qubesdb make the needed `mkdir` and `setxattr` calls manually.  This has the advantage that we have full control over everything involved, rather than relying on external code (SELinux policy and libraries, GNU coreutils) that might or might not be working at this stage.  It also works even if systemd is not in use, as will be the case on Alpine.  The only drawbacks are that everything needs to be done in C and that policies that use different labels from the ones we provide will not work.  The latter can be fixed by allowing the label to be overridden with a command-line argument.

**Comment by marmarek** [comment]
Can you find out what/why fails first, before trying to find a solution? Since it fails even without enabling selinux, it's probably enough to install just this one package in a test vm. 

Note the unit file (and the service itself) must also work without selinux (not only disabled, but also policy not even installed). I think the original `mkdir -pZ` call was correct and the problem was elsewhere (default file contexts list?). Maybe even plain `mkdir -p` should be enough, with proper default contexts set. Most services work just fine with just proper policy, without any special handling for setting file contexts in the services themselves, I don't see why qubes-db should be any different.

**Comment by DemiMarie** [comment]
> Can you find out what/why fails first, before trying to find a solution? Since it fails even without enabling selinux, it's probably enough to install just this one package in a test vm.
> 
> Note the unit file (and the service itself) must also work without selinux (not only disabled, but also policy not even installed). I think the original `mkdir -pZ` call was correct and the problem was elsewhere (default file contexts list?). Maybe even plain `mkdir -p` should be enough, with proper default contexts set. Most services work just fine with just proper policy, without any special handling for setting file contexts in the services themselves, I don't see why qubes-db should be any different.

Where the initial tests done with _all_ of the SELinux policies installed, or with only some of them?  All of my tests included all of the policy packages, and I would not be surprised if there were dependencies between them that I had not considered.  My personal SELinux-enabled qubes (with a policy roughly equivalent to all of the policies combined) has worked reliably for well over 2 years, and a test with all of the packages installed also passed.

**Comment by marmarek** [comment]
The failure above was before installing any extra package. It's just `dnf update` on a default template with the build result included as an extra repository. I believe you can trivially achieve sufficiently similar (broken) state by taking default template and doing `dnf update ./qubes-db*rpm`.

**Comment by DemiMarie** [comment]
> The failure above was before installing any extra package. It's just `dnf update` on a default template with the build result included as an extra repository. I believe you can trivially achieve sufficiently similar (broken) state by taking default template and doing `dnf update ./qubes-db*rpm`.

What about the failure with SELinux enforcing?  This module simply does not work if `qubes-core-agent-selinux` is not installed, and in fact even trying to build it without that package is somewhat questionable.  I will add `Requires:` and `BuildRequires:` to indicate this.

**Comment by marmarek** [comment]
> What about the failure with SELinux enforcing? 

Not sure, but when I tried `mkdir -pZ` in various directories on a (non-Qubes) SELinux-enabled Fedora, it worked as expected:
```
# ls -ldZ /run/samba
ls: cannot access '/run/samba': No such file or directory
# mkdir -pZ /run/samba
# ls -ldZ /run/samba
drwxr-xr-x. 2 root root unconfined_u:object_r:smbd_var_run_t:s0 40 Jul 18 15:36 /run/samba
# grep -rn /run/samba  /etc/selinux/
/etc/selinux/targeted/contexts/files/file_contexts:1232:/var/run/samba(/.*)?	system_u:object_r:smbd_var_run_t:s0
...
```

>  I will add Requires:

To the `-selinux` subpackage, makes sense. As for `BuildRequires:` - what would it be needed for? Currently it builds just fine without it.

**Comment by DemiMarie** [comment]
> > I will add Requires:
> 
> To the `-selinux` subpackage, makes sense. As for `BuildRequires:` - what would it be needed for? Currently it builds just fine without it.

`qubes-db-vm-selinux` only builds without `qubes-core-agent-selinux` because of an `optional` block, but that block also hides what should be build-time failures and turns them into run-time failures that are much harder to debug.  I would much rather get a build-time error, but that means that `qubes-core-agent-selinux` will need to be installed for `qubes-db-vm-selinux` to build.  I believe the same problem applies to all of the other SELinux policy packages.

**Comment by marmarek** [comment]
Such BR will be problematic, as it might introduce circular dependency, which are especially bad during bootstrapping (including initial build of selinux-enabled packages). core-agent-linux logically depends on core-qubesdb. There is at least runtime explicit dependency, I don't see explicit build time dependency now, but there might be indirect one (and even if it isn't now, it could be added later when needed).

**Comment by DemiMarie** [comment]
> Such BR will be problematic, as it might introduce circular dependency, which are especially bad during bootstrapping (including initial build of selinux-enabled packages). core-agent-linux logically depends on core-qubesdb. There is at least runtime explicit dependency, I don't see explicit build time dependency now, but there might be indirect one (and even if it isn't now, it could be added later when needed).

Okay, I dropped the BR.  The runtime dependency is still there, though.  It should not result in a circular dependency, as `qubes-core-agent-selinux` does _not_ depend on `qubes-core-agent`.

**Review by marmarek — APPROVED** [review]

**Pull request merged**